File: packages/python-sdk/workflow_engine/webhooks.py
import hmac
import hashlib


def verify_webhook_signature(payload: str, signature: str, secret: str) -> bool:
    """
    Verify HMAC-SHA256 signature for incoming webhooks.

    :param payload: Raw string body of the incoming request
    :param signature: Value from the X-Webhook-Signature header
    :param secret: Webhook signing secret
    :return: True if signature is valid, False otherwise
    """
    if not signature or not payload or not secret:
        return False

    try:
        expected = hmac.new(
            secret.encode('utf-8'),
            payload.encode('utf-8'),
            hashlib.sha256
        ).hexdigest()

        return hmac.compare_digest(signature, expected)
    except Exception:
        return False


def generate_webhook_signature(payload: str, secret: str) -> str:
    """Generate HMAC-SHA256 signature for testing"""
    return hmac.new(
        secret.encode('utf-8'),
        payload.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()

File: packages/javascript-sdk/src/webhooks.ts
import * as crypto from 'crypto';

/**
 * Verify the HMAC signature of a webhook payload
 * @param payload - The raw request body as a string
 * @param signature - The signature from the X-Webhook-Signature header
 * @param secret - The webhook secret
 * @returns true if the signature is valid
 */
export function verifyWebhookSignature(
  payload: string,
  signature: string | undefined,
  secret: string,
): boolean {
  if (!signature) {
    return false;
  }

  try {
    const hmac = crypto.createHmac('sha256', secret);
    hmac.update(payload);
    const expectedSignature = hmac.digest('hex');

    // Use timing-safe comparison
    return crypto.timingSafeEqual(
      Buffer.from(signature),
      Buffer.from(expectedSignature)
    );
  } catch {
    return false;
  }
}

/**
 * Generate a webhook signature for testing
 * @param payload - The payload to sign
 * @param secret - The webhook secret
 * @returns HMAC signature
 */
export function generateWebhookSignature(payload: string, secret: string): string {
  const hmac = crypto.createHmac('sha256', secret);
  hmac.update(payload);
  return hmac.digest('hex');
}
